AdminServer/appscale/admin/instance_manager/projects_manager.py

```diff
@@ -0,0 +1,254 @@
+          self.callback)
+
+      self[version_id].ensure_watch()
+
+  def stop(self):
+    """ Stops all watches associated with this service. """
+    for version_id in self:
+      del self[version_id]
+
+    self._stopped = True
+
+  def _update_versions_watch(self, new_versions_list):
+    """ Handles the creation and deletion of a service's versions.
+
+    Since this runs in a separate thread, it doesn't change any state directly.
+    Instead, it just acts as a bridge back to the main IO loop.
+
+    Args:
+      new_versions_list: A fresh list of strings specifying a service's
+        version IDs.
+    """
+    if self._stopped:
+      return False
+
+    main_io_loop = IOLoop.instance()
+    main_io_loop.add_callback(self.update_versions, new_versions_list)
+
+
+class ProjectManager(dict):
+  """ Keeps track of services for a project. """
+  def __init__(self, zk_client, project_id, callback):
+    """ Creates a new ProjectManager.
+
+    Args:
+      zk_client: A KazooClient.
+      project_id: A string specifying a project ID.
+      callback: A function to call whenever a version is updated.
+    """
+    super(ProjectManager, self).__init__()
+    self.zk_client = zk_client
+    self.project_id = project_id
+    self.callback = callback
+    self._stopped = False
+
+    self.services_node = '/appscale/projects/{}/services'.format(project_id)
+    zk_client.ensure_path(self.services_node)
+    self.watch = zk_client.ChildrenWatch(self.services_node,
+                                         self._update_services_watch)
+
+  def update_services(self, new_services_list):
+    """ Establishes watches for all of a project's services.
+
+    Args:
+      new_services_list: A fresh list of strings specifying a project's
+        service IDs.
+    """
+    to_stop = [service for service in self if service not in new_services_list]
+    for service_id in to_stop:
+      self[service_id].stop()
+      del self[service_id]
+
+    for service_id in new_services_list:
+      if service_id not in self:
+        self[service_id] = ServiceManager(self.zk_client, self.project_id,
+                                          service_id, self.callback)
+
+  def stop(self):
+    """ Stops all watches associated with this service. """
+    for service_id in self:
+      self[service_id].stop()
+      del self[service_id]
+
+    self._stopped = True
+
+  def _update_services_watch(self, new_services_list):
+    """ Handles the creation and deletion of a project's services.
+
+    Since this runs in a separate thread, it doesn't change any state directly.
+    Instead, it just acts as a bridge back to the main IO loop.
+
+    Args:
+      new_services_list: A fresh list of strings specifying a project's
+        service IDs.
+    """
+    if self._stopped:
+      return False
+
+    main_io_loop = IOLoop.instance()
+    main_io_loop.add_callback(self.update_services, new_services_list)
+
+
+class ProjectsManager(dict):
```

Comment:
can this be changed to `GlobalProjectsManager` it seems to easy too confuse with the ProjectManager